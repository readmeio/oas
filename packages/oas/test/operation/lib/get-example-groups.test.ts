import { beforeAll, test, expect } from 'vitest';

import Oas from '../../../src/index.js';

let readmeExtensions: Oas;
let requestExamples: Oas;
let trainTravel: Oas;

beforeAll(async () => {
  readmeExtensions = await import('@readme/oas-examples/3.0/json/readme-extensions.json')
    .then(r => r.default)
    .then(Oas.init);
  await readmeExtensions.dereference();

  requestExamples = await import('@readme/oas-examples/3.0/json/request-examples.json')
    .then(r => r.default)
    .then(Oas.init);
  await requestExamples.dereference();

  trainTravel = await import('@readme/oas-examples/3.1/json/train-travel.json').then(r => r.default).then(Oas.init);
  await trainTravel.dereference();
});

test('body/header/path/query param examples with matching response examples', () => {
  const operation = requestExamples.operation('/parameterExamples/{param1}/{param2}', 'patch');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('body param examples with matching response examples', () => {
  const operation = trainTravel.operation('/bookings/{bookingId}/payment', 'post');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('body param examples with matching response examples (primitive)', () => {
  const operation = requestExamples.operation('/requestBody-primitive-example', 'patch');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
  expect(groups.cat.request.body).toBeTypeOf('string');
  expect(groups.cat.response.mediaTypeExample.value).toBeTypeOf('string');
});

test('path param examples with matching response examples', () => {
  const operation = requestExamples.operation('/parameterExamples/{param1}/{param2}', 'get');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('form-urlencoded params with matching response example', () => {
  const operation = requestExamples.operation('/requestBody-form-data-example', 'post');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('custom code samples with matching response examples', () => {
  const operation = readmeExtensions.operation('/x-code-samples', 'post');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('custom code samples with no matching response examples', () => {
  const operation = readmeExtensions.operation('/x-code-samples', 'get');
  const groups = operation.getExampleGroups();
  expect(groups).toMatchSnapshot();
});

test('body param example with no title to match responses against', () => {
  const operation = trainTravel.operation('/bookings', 'post');
  const groups = operation.getExampleGroups();
  expect(groups).toStrictEqual({});
});

test('invalid operation', () => {
  const operation = trainTravel.operation('/invalid', 'patch');
  const groups = operation.getExampleGroups();
  expect(groups).toStrictEqual({});
});
