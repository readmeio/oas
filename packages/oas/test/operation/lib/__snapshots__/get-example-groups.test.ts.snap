// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`body param examples with matching response examples (primitive) 1`] = `
{
  "cat": {
    "name": "An example of a cat",
    "request": {
      "body": "{
  "name": "Fluffy",
  "petType": "Cat"
}",
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "An example of a cat",
        "title": "cat",
        "value": "{
  "name": "Fluffy",
  "petType": "Cat"
}",
      },
      "status": "200",
    },
  },
  "dog": {
    "name": "An example of a dog",
    "request": {
      "body": "{
  "name": "Buster",
  "petType": "Dog"
}",
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "An example of a dog",
        "title": "dog",
        "value": "{
  "name": "Buster",
  "petType": "Dog"
}",
      },
      "status": "200",
    },
  },
}
`;

exports[`body param examples with matching response examples 1`] = `
{
  "Bank": {
    "name": "Bank Account Payment",
    "request": {
      "body": {
        "amount": 100.5,
        "currency": "gbp",
        "source": {
          "account_type": "individual",
          "bank_name": "Starling Bank",
          "country": "gb",
          "name": "J. Doe",
          "number": "00012345",
          "object": "bank_account",
          "sort_code": "000123",
        },
      },
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "Bank Account Payment",
        "title": "Bank",
        "value": {
          "amount": 100.5,
          "currency": "gbp",
          "id": "2e3b4f5a-6b7c-8d9e-0f1a-2b3c4d5e6f7a",
          "links": {
            "booking": "https://api.example.com/bookings/1725ff48-ab45-4bb5-9d02-88745177dedb",
          },
          "source": {
            "account_type": "individual",
            "bank_name": "Starling Bank",
            "country": "gb",
            "name": "J. Doe",
            "number": "*********2345",
            "object": "bank_account",
            "sort_code": "000123",
          },
          "status": "succeeded",
        },
      },
      "status": "200",
    },
  },
  "Card": {
    "name": "Card Payment",
    "request": {
      "body": {
        "amount": 49.99,
        "currency": "gbp",
        "source": {
          "address_city": "London",
          "address_country": "gb",
          "address_line1": "123 Fake Street",
          "address_line2": "4th Floor",
          "address_post_code": "N12 9XX",
          "cvc": 123,
          "exp_month": 12,
          "exp_year": 2025,
          "name": "J. Doe",
          "number": "4242424242424242",
          "object": "card",
        },
      },
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "Card Payment",
        "title": "Card",
        "value": {
          "amount": 49.99,
          "currency": "gbp",
          "id": "2e3b4f5a-6b7c-8d9e-0f1a-2b3c4d5e6f7a",
          "links": {
            "booking": "https://api.example.com/bookings/1725ff48-ab45-4bb5-9d02-88745177dedb/payment",
          },
          "source": {
            "address_country": "gb",
            "address_post_code": "N12 9XX",
            "cvc": 123,
            "exp_month": 12,
            "exp_year": 2025,
            "name": "J. Doe",
            "number": "************4242",
            "object": "card",
          },
          "status": "succeeded",
        },
      },
      "status": "200",
    },
  },
}
`;

exports[`body/header/path/query param examples with matching response examples 1`] = `
{
  "userRegistration": {
    "name": "example summary (param 1)",
    "request": {
      "body": {
        "email": "test@example.com",
        "id": 1234,
        "name": "Test user name",
      },
      "header": {
        "param5": "param5-example",
      },
      "path": {
        "param1": "param1-example",
        "param2": "param2-example",
      },
      "query": {
        "param3": "param3-example",
      },
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "userRegistration",
        "title": "userRegistration",
        "value": {
          "description": "a lengthier example description (response)",
          "email": "test+response@example.com",
          "id": 1234,
          "name": "Test user name (response)",
          "summary": "example summary (response)",
        },
      },
      "status": "200",
    },
  },
}
`;

exports[`custom code samples with matching response examples 1`] = `
{
  "NoCorrespondingResponseForCustomCodeSample": {
    "customCodeSamples": [
      {
        "code": "# This custom cURL snippet does not have a custom name so it has the name of "Default #3".

curl -X POST https://api.example.com/v2/alert",
        "language": "curl",
        "name": "Default #3",
      },
    ],
    "name": "Default #3",
  },
  "TestExample": {
    "customCodeSamples": [
      {
        "code": "curl -X POST https://api.example.com/v2/alert",
        "correspondingExample": "TestExample",
        "install": "brew install curl",
        "language": "curl",
        "name": "Custom cURL snippet",
      },
      {
        "code": "curl -X POST https://api.example.com/v2/alert",
        "correspondingExample": "TestExample",
        "install": "brew install curl",
        "language": "curl",
        "name": "Yet another custom snippet",
      },
      {
        "code": "import requests

print("do something idk")",
        "correspondingExample": "TestExample",
        "install": "pip install requests",
        "language": "python",
        "name": "Yet another custom snippet",
      },
    ],
    "name": "Custom cURL snippet",
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "An example of a cat",
        "title": "TestExample",
        "value": {
          "breed": "Persian",
          "color": "White",
          "gender": "male",
          "name": "Fluffy",
          "petType": "Cat",
        },
      },
      "status": "200",
    },
  },
  "curlExample": {
    "customCodeSamples": [
      {
        "code": "http POST https://api.example.com/v2/another-alert",
        "correspondingExample": "curlExample",
        "install": "brew install httpie",
        "language": "curl",
        "name": "Another cURL snippet",
      },
      {
        "code": "import something from 'requestidk';

console.log('do something idk');",
        "correspondingExample": "curlExample",
        "install": "npm install requestidk",
        "language": "node",
        "name": "Yet another custom snippet",
      },
    ],
    "name": "Another cURL snippet",
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "An example of a cat",
        "title": "curlExample",
        "value": {
          "missing": "can't find your stuff dawg",
        },
      },
      "status": "404",
    },
  },
}
`;

exports[`custom code samples with no matching response examples 1`] = `
{
  "NoCorrespondingResponseForCustomCodeSample": {
    "customCodeSamples": [
      {
        "code": "curl -X POST https://api.example.com/v2/alert",
        "install": "brew install curl",
        "language": "curl",
        "name": "Custom cURL snippet",
      },
      {
        "code": "# This custom cURL snippet does not have a custom name so it has the name of "Default #2".

curl -X POST https://api.example.com/v2/alert",
        "language": "curl",
        "name": "Default #2",
      },
    ],
    "name": "Custom cURL snippet",
  },
}
`;

exports[`form-urlencoded params with matching response example 1`] = `
{
  "auth_example": {
    "name": "auth_example",
    "request": {
      "formData": {
        "client_id": "id123",
        "client_secret": "secret456",
        "scope": 789,
      },
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "auth_example",
        "title": "auth_example",
        "value": {
          "access_token": 123,
          "expires_in": 789,
          "token_type": "type456",
        },
      },
      "status": "200",
    },
  },
}
`;

exports[`path param examples with matching response examples 1`] = `
{
  "userRegistration": {
    "name": "example summary (param 1)",
    "request": {
      "path": {
        "param1": "param1-example",
        "param2": "param2-example",
      },
    },
    "response": {
      "mediaType": "application/json",
      "mediaTypeExample": {
        "summary": "userRegistration",
        "title": "userRegistration",
        "value": {
          "description": "a lengthier example description (response)",
          "email": "test+response@example.com",
          "id": 1234,
          "name": "Test user name (response)",
          "summary": "example summary (response)",
        },
      },
      "status": "200",
    },
  },
}
`;
